doc/design/publish-1.md

```diff
@@ -0,0 +1,141 @@
+# Moving Parts
+
+This is a summary of system components that are relevant to publish and how they relate to eachother.
+
+### Dandiset
+A DANDI dataset.
+Each dandiset is guaranteed to have a draft Version, and can also have any number of published Versions.
+
+### Version
+Versions consist of a JSON blob of metadata and any number of Assets.
+There are two kinds of Version: draft and published.
+Published Versions are immutable, only draft Versions are editable (except admins, who have superpowers).
+Versions also track their validation status.
+
+### AssetBlob
+This is basically just a pointer to a blob in S3.
+AssetBlobs are central to how we implement deduplication of blobs in S3.
+Multiple Assets can refer to the same AssetBlob, meaning we only need to store each blob in S3 once.
+
+Whenever a new AssetBlob is created, a background task runs to compute the sha256 checksum (a "digest") of the blob.
+This can take a long time for large blobs, as it must stream all of the data from S3.
+
+### Asset
+Assets consist of a JSON blob of metadata and a reference to an AssetBlob.
+Versions and Assets are many-to-many; a Version can have any number of Assets, and an Asset can belong to any number of Versions (even, hypothetically, versions in different Dandisets).
+Assets also track their validation status.
+
+Similarly to Versions, there are two kinds of Asset: draft and published.
+Draft Assets are only allowed to belong to draft Versions.
+Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).
+
+Note that published Assets can belong to both draft and published Versions.
+
+Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
+Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
+This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.
```

Comment:
I would reorder these two paragraphs to go from basic "Asset" behavior to IMHO yet to be finalized "published vs draft" Asset fresh discussion:
```suggestion
Uploading new data or modifying the metadata of an Asset doesn't actually modify the Asset.
Instead, it creates a new Asset and replaces the reference to that Asset in the Version.
This ensures that other Versions which might refer to the original Asset are unmodified, while the Version containing the Asset being modified reflects the changes correctly.

Similarly to Versions, there are two kinds of Asset: draft and published.
Draft Assets are only allowed to belong to draft Versions.
Published Assets are basically the same as draft Assets, but have some extra metadata included that was defined during the publish process (`datePublished`, `publishedBy`, etc).

Note that published Assets can belong to both draft and published Versions.
```